### PR TITLE
Trusted claims

### PR DESCRIPTION
**Pull request by traiansf**
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

You should probably review the commits separately, as the `Refactoring IndexedModule to include axiom attrs` touches a lot of code (and does not do much).

Also, it will probably be untestable until https://github.com/kframework/k/pull/225 is merged

### CODE REVIEW

File: src/main/haskell/kore/src/Kore/Attribute/Trusted.hs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{-|
+Module      : Kore.Attribute.Trusted
+Description : Trusted claim sentence attribute
+Copyright   : (c) Runtime Verification, 2018
+License     : NCSA
+Maintainer  : thomas.tuegel@runtimeverification.com
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Attribute/Trusted.hs]: Why aren't you the maintainer?

File: src/main/haskell/kore/src/Kore/Attribute/Trusted.hs [hunk 2]
```diff
@@ -0,0 +1,60 @@
+{-|
+Module      : Kore.Attribute.Trusted
+Description : Trusted claim sentence attribute
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Attribute/Trusted.hs]: How about "Attribute identifying trusted claim sentences"?

Regardless of whether you like my suggestion or not, could explain in a few words what is a "trusted claim"?

File: src/main/haskell/kore/src/Kore/Exec.hs [hunk 3]
```diff
@@ -348,24 +349,26 @@ prove limit definitionModule specModule = do
         (rewriteAxioms, simplifier, substitutionSimplifier) =
             axiomsAndSimplifiers
     specAxioms <-
-        simplifyRewriteAxioms tools
+        mapM (sequence . fmap (simplifyRewriteRule tools))
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Exec.hs]: Could you extract the "sequence . fmap (simplifyRewriteRule tools)" or something similar to a separate function that has a signature?

File: src/main/haskell/kore/src/Kore/Exec.hs [hunk 4]
```diff
@@ -348,24 +349,26 @@ prove limit definitionModule specModule = do
...
 
     return $
         Bifunctor.first
             (give symbolOrAlias toMLPattern)
             result
 
   where
-    makeClaim claim = (claim, limit)
+    makeClaim (attributes, rule) = Claim { rule , attributes }
+    extractUntrustedClaims =
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Exec.hs]: Could you add a signature for this function?

File: src/main/haskell/kore/src/Kore/IndexedModule/IndexedModule.hs [hunk 5]
```diff
@@ -123,14 +126,60 @@ data IndexedModule param pat atts =
         -- ^ map from builtin domain (symbol and sort) identifiers to the hooked
         -- identifiers
     }
-    deriving (Functor, Generic, Show)
+    deriving (Generic, Show)
+
+instance Bifunctor (IndexedModule param pat) where
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/IndexedModule/IndexedModule.hs]: Is this really a good idea? I mean, when someone is reading the code, why would it make sense that "first" and "second" for an indexed module would refer to a rather obscure part of it (i.e. one kind of attributes)?

I know that those are the last two arguments, and Bifunctor works that way, but does this actually make sense, or does it look more like a hack that saves a bit of work?

Assuming that my questions make some sense, then why not make it a Trifunctor? That way you're not making it a functor over the obscure things only, you're making it a functor over (almost) everything.

My questions are genuine, not rethorical.

File: src/main/haskell/kore/test/Test/Kore/OnePath/Verification.hs [hunk 6]
```diff
@@ -262,6 +265,42 @@ test_onePathVerification = give symbolOrAliasSorts
         assertEqualWithExplanation ""
             (Left $ ExpandedPattern.fromPurePattern Mock.b)
             actual
+    , testCase "trusted second proves first" $ do
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/test/Test/Kore/OnePath/Verification.hs]: Could you also add a test that trusted claims are not used in the first step? E.g. from
trusted claim: a=>b
claim: a=>b
you get a failure?

File: src/main/haskell/kore/src/Kore/IndexedModule/IndexedModule.hs (hunk 5 above)
Comment (traiansf) [reply src/main/haskell/kore/src/Kore/IndexedModule/IndexedModule.hs]: Well, I'm not sure.  I was just continuing existing work which made it a `Functor`.  On the other hand, the only usage is to replace the attributes by `Null`.  I guess it made more sense when it was just a `Functor` and the instance was automatically derived.

Comment (vladciobanu) [reply src/main/haskell/kore/src/Kore/IndexedModule/IndexedModule.hs]: I also found that bit weird even from before (that `IndexedModule` is a `Functor` for its attributes. Perhaps lenses would be a better choice here?

Comment (traiansf) [reply src/main/haskell/kore/src/Kore/IndexedModule/IndexedModule.hs]: I just went on an removed teh Bifunctor instance. If we ever need it for more than one function, we can revisit this issue.

**Comment by traiansf** [comment]
@virgil-serbanuta are you satisfied with the changes?

**Comment by dwightguth** [comment]
Do we really want to replicate the trusted attribute of the java backend? wouldn't it be better to simply tell the prover when you prove which claims you want to prove and which claims you want to trust?

**Comment by traiansf** [comment]
@dwightguth I don't have anything against using a different mechanism, but that would have to be designed and work from the frontend towards the backend.  We could separate them in different modules and send two such modules to the backend, but I'm not sure it would be such a great improvement. 

Also, languages such as Coq sometimes use something like `Admitted` to accept a result without proof.

**Comment by traiansf** [comment]
Jenkins: test this please

**Pull request merged**